apollo-router-core/src/spec/query.rs

```diff
@@ -595,85 +840,1527 @@ mod tests {
+            "query($foo:Foo){x}",
+            json!({})
+        );
+        assert_validation!(
+            "type Foo{} type Query { x: String }",
+            "query($foo:Foo){x}",
+            json!({"foo":{}})
+        );
+        assert_validation_error!(
+            "type Foo{} type Query { x: String }",
+            "query($foo:Foo){x}",
+            json!({"foo":1})
+        );
+        assert_validation_error!(
+            "type Foo{} type Query { x: String }",
+            "query($foo:Foo){x}",
+            json!({"foo":"str"})
+        );
+        assert_validation_error!(
+            "type Foo{x:Int!} type Query { x: String }",
+            "query($foo:Foo){x}",
+            json!({"foo":{}})
+        );
         assert_validation!(
-            "type Foo{x:Int!}",
+            "type Foo{x:Int!} type Query { x: String }",
             "query($foo:Foo){x}",
             json!({"foo":{"x":1}})
         );
         assert_validation!(
-            "type Foo implements Bar interface Bar{x:Int!}",
+            "type Foo implements Bar interface Bar{x:Int!} type Query { x: String }",
             "query($foo:Foo){x}",
             json!({"foo":{"x":1}}),
         );
         assert_validation_error!(
-            "type Foo implements Bar interface Bar{x:Int!}",
+            "type Foo implements Bar interface Bar{x:Int!} type Query { x: String }",
             "query($foo:Foo){x}",
             json!({"foo":{"x":"str"}}),
         );
         assert_validation_error!(
-            "type Foo implements Bar interface Bar{x:Int!}",
+            "type Foo implements Bar interface Bar{x:Int!} type Query { x: String }",
             "query($foo:Foo){x}",
             json!({"foo":{}}),
         );
-        assert_validation!("scalar Foo", "query($foo:Foo!){x}", json!({"foo":{}}));
-        assert_validation!("scalar Foo", "query($foo:Foo!){x}", json!({"foo":1}));
-        assert_validation_error!("scalar Foo", "query($foo:Foo!){x}", json!({}));
         assert_validation!(
-            "type Foo{bar:Bar!} type Bar{x:Int!}",
+            "scalar Foo type Query { x: String }",
+            "query($foo:Foo!){x}",
+            json!({"foo":{}})
+        );
+        assert_validation!(
+            "scalar Foo type Query { x: String }",
+            "query($foo:Foo!){x}",
+            json!({"foo":1})
+        );
+        assert_validation_error!(
+            "scalar Foo type Query { x: String }",
+            "query($foo:Foo!){x}",
+            json!({})
+        );
+        assert_validation!(
+            "type Foo{bar:Bar!} type Bar{x:Int!} type Query { x: String }",
             "query($foo:Foo){x}",
             json!({"foo":{"bar":{"x":1}}})
         );
     }
+
+    #[test]
+    fn filter_root_errors() {
+        let schema = "type Query {
+            getInt: Int
+            getNonNullString: String!
+        }";
+        let query = "query MyOperation { getInt }";
+        let response = json! {{
+            "getInt": "1",
```

Comment:
It took me a while to notice the data coming back was not an int. For all the test failures it could be worth putting a comment as to why the result is expected with links to the spec if available.

Reply:
right, good idea

Reply:
comments in [2f7b76c](https://github.com/apollographql/router/pull/528/commits/2f7b76cf0f7f795155ee155f12a5e07d989b5440)